### PR TITLE
[update.sh] Parsing token inside each upstream

### PR DESCRIPTION
**Pull request by sazriel26**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

This pull request is not for formula but for update.sh script to allow local token parsing inside each upstream

-----

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
@sazriel26 can you explain the motivation for this PR rather than what it is doing? Thanks!

**Comment by MikeMcQuaid** [comment]
@sazriel26 Ok, I understand a bit more about what you're doing here but not the "why" behind the pull request. What did it do before that you didn't like and what does it do now instead?

File: Library/Homebrew/cmd/update.sh [hunk 1]
```diff
@@ -513,12 +513,12 @@ EOS
 
   # HOMEBREW_GITHUB_API_TOKEN is optionally defined in the user environment.
   # shellcheck disable=SC2153
-  if [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
-  then
-    CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${HOMEBREW_GITHUB_API_TOKEN}")
-  else
-    CURL_GITHUB_API_ARGS=()
-  fi
+  #if [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:516]: Either delete or uncomment this

File: Library/Homebrew/cmd/update.sh [hunk 2]
```diff
@@ -637,12 +637,36 @@ EOS
...
       UPSTREAM_REPOSITORY_URL="$(git config remote.origin.url)"
+      UPSTREAM_REPOSITORY_TOPARSE="${UPSTREAM_REPOSITORY_URL#*://}"
+      UPSTREAM_REPOSITORY_L1="${UPSTREAM_REPOSITORY_TOPARSE%%/*}"
+      UPSTREAM_REPOSITORY_DOMAIN="${UPSTREAM_REPOSITORY_L1#*@}"
+      [[ "${UPSTREAM_REPOSITORY_L1}" == "${UPSTREAM_REPOSITORY_DOMAIN}" ]] \
+        && UPSTREAM_REPOSITORY_USERPASS="" \
+        || UPSTREAM_REPOSITORY_USERPASS="${UPSTREAM_REPOSITORY_L1%@*}"
+      [[ -z "${UPSTREAM_REPOSITORY_USERPASS%:*}" ]] || [[ "${UPSTREAM_REPOSITORY_USERPASS%:*}" == "git" ]] \
+        && UPSTREAM_REPOSITORY_TOKEN="" \
+        || UPSTREAM_REPOSITORY_TOKEN="${UPSTREAM_REPOSITORY_USERPASS#*:}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:648]: This is very hard to follow what is going on. Can you add some example URL formats you're parsing here? I also think ideally this would be calling out to something in `git` if possible rather than parsing here directly and perhaps moving this logic to another function for readability.

File: Library/Homebrew/cmd/update.sh [hunk 3]
```diff
@@ -637,12 +637,36 @@ EOS
...
+      fi
 
       # HOMEBREW_UPDATE_FORCE and HOMEBREW_UPDATE_AUTO aren't modified here so ignore subshell warning.
       # shellcheck disable=SC2030
-      if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]]
+      if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]] \
+      || [[ "${UPSTREAM_REPOSITORY_DOMAIN%:*}" == "github.com" ]]
       then
-        UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY_URL#https://github.com/}"
+        #UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY_URL#https://github.com/}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:668]: Please delete this or comment it out

**Review by MikeMcQuaid — COMMENTED** [review]
I understand a bit more what you're asking for here, thanks

File: Library/Homebrew/cmd/update.sh [hunk 4]
```diff
@@ -637,13 +629,32 @@ EOS
...
       UPSTREAM_REPOSITORY_URL="$(git config remote.origin.url)"
+      eval UPSTREAM_REPOSITORY_URL_PARSED=( $(url_get "${UPSTREAM_REPOSITORY_URL}" user pass host path) )
+      if [[ ${#UPSTREAM_REPOSITORY_URL_PARSED[@]} -ne 4 ]]; then
+        echo "Failed to parse repository URL=${UPSTREAM_REPOSITORY_URL} \!" >&2
+        exit
+      fi
+
+
+      # HOMEBREW_GITHUB_API_TOKEN is optionally defined in the user environment.
+      # can be supersede by local repository URL
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:640]:
```suggestion
      # can be superseded a token from the local repository URL
```

File: Library/Homebrew/utils/url.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# url_get <URL> [key] [key] ...
+# where:
+#   URL ~ [scheme://][user[:pass]@]host[:port][path]
+#   key is one of the element composing the URL
+url_get() {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/url.sh:5]: This can still live in `update.sh`

File: Library/Homebrew/utils/url.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  [[ ${#} -gt 0 ]] || return
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/url.sh:6]: This shouldn't fail silently.

File: Library/Homebrew/utils/url.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  local url="${1}"
+  [[ -z "${url}" ]] && return
+  shift
+
+  local _uphpp="${url#*://}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/url.sh:12]: No idea what `uphpp` means. Please use longer variable names.

File: Library/Homebrew/utils/url.sh [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  local _scheme="${url%%://*}"
+  if [[ "${url}" == "${_uphpp}" ]]; then
+    _scheme=""
+  fi
+
+  local _hpp="${_uphpp#*@}"
+  local _up="${_uphpp%%@*}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/url.sh:19]: No idea what these variable names mean

File: Library/Homebrew/utils/url.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  if [[ "${_hp}" == "${_hpp}" ]]; then
+    _path=""
+  elif [[ -z "${_path}" ]]; then
+    _path="/"
+  fi
+
+  local _host="${_hp%:*}"
+  local _port="${_hp#*:}"
+  if [[ "${_hp}" == "${_host}" ]]; then
+    _port=""
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/url.sh:43]: What is the port used for here?

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the PR! Just to set expectations: I think this feature is desirable but this PR is currently quite a long way off being an implementation we can merge and support. I find it very hard to read and understand right now and the error-handling in `update.sh` needs to be absolutely rock-solid otherwise we run a risk of breaking Homebrew's own update mechanism.

Thanks again!

**Comment by MikeMcQuaid** [comment]
> First thank you @MikeMcQuaid for your kind support and availability about this PR.

You're welcome ❤️ 



> According to [#16289 (review)](https://github.com/Homebrew/brew/pull/16289#pullrequestreview-1768569979), it looks hard to satisfy some prerequisites in contributing for a newcomer like me, isn't it?

I think we can get there together; I'm just giving you the heads up that it may take a while.

Some newcomers contributions are merged very quickly, others are not. It depends on the readability of the code, the invasiveness of the change and the potential negative effects of any issues.

> What about to discard this PR which is not ready at this moment for merge?

Up to you, I'm happy to keep working here if you are!



> Thanks again for your understanding and sorry for the annoyance (including time consuming)

Thank you and no apologies needed; not annoyed here at all 😁

File: Library/Homebrew/cmd/update.sh [hunk 10]
```diff
@@ -640,10 +631,21 @@ EOS
 
       # HOMEBREW_UPDATE_FORCE and HOMEBREW_UPDATE_AUTO aren't modified here so ignore subshell warning.
       # shellcheck disable=SC2030
-      if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]]
+      if [[ "${UPSTREAM_REPOSITORY_URL}" =~ https://(([^:@]+)(:([^@]+))?@)?github.com/(.*)$ ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:634]:
```suggestion
      if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]] || 
         [[ "${UPSTREAM_REPOSITORY_URL}" =~ https://(([^:@]+)(:([^@]+))?@)?github.com/(.*)$ ]]
```

This may allow the regex to be made simpler and ensures the previous logic takes priority.

File: Library/Homebrew/cmd/update.sh [hunk 11]
```diff
@@ -640,10 +631,21 @@ EOS
...
+        # shellcheck disable=SC2153
+        if [[ -n "${BASH_REMATCH[4]}" ]]
+        then
+          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${BASH_REMATCH[4]}")
+        elif [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
+        then
+          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${HOMEBREW_GITHUB_API_TOKEN}")
+        else
+          CURL_GITHUB_API_ARGS=()
+        fi
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:648]:
```suggestion
        # HOMEBREW_GITHUB_API_TOKEN is optionally defined in the user environment.
        # can be superseded a token from the local repository URL
        # shellcheck disable=SC2153
        if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]] && 
           [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
        then
          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${HOMEBREW_GITHUB_API_TOKEN}")
        elif [[ -n "${BASH_REMATCH[4]}" ]]
        then
          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${BASH_REMATCH[4]}")
        else
          CURL_GITHUB_API_ARGS=()
        fi
```

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks @sazriel26. This is looking a lot clearer now. I've suggested some tweaks (that may need fixed up with `brew style`) to ensure your new logic is a fallback rather than taking priority or encompassing the old logic.

File: Library/Homebrew/cmd/update.sh (hunk 10 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/update.sh:634]: @sazriel26 This change didn't seem to be made. If there's a token in the URL: it will never match `"https://github.com/"*` so that conditional should always be present first.

File: Library/Homebrew/cmd/update.sh (hunk 11 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/update.sh:648]: @sazriel26 This change also doesn't seem to have been made. It's important that your new logic is a fallback to the existing logic, not a replacement. I do not trust it to behave consistently and any issues with this file mean people cannot get updates from Homebrew.

File: Library/Homebrew/cmd/update.sh [hunk 12]
```diff
@@ -640,10 +631,31 @@ EOS
 
       # HOMEBREW_UPDATE_FORCE and HOMEBREW_UPDATE_AUTO aren't modified here so ignore subshell warning.
       # shellcheck disable=SC2030
-      if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]]
+      if [[ "${UPSTREAM_REPOSITORY_URL}" = "https://github.com/"* ]] ||
+         [[ "${UPSTREAM_REPOSITORY_URL}" =~ https://(([^:@]+)(:([^@]+))?@)?github.com/(.*)$ ]]
       then
-        UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY_URL#https://github.com/}"
-        UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY%.git}"
+        if [[ -n "${BASH_REMATCH[5]}" ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:637]: Again: this should be not be checked first if `[[ "${UPSTREAM_REPOSITORY_URL}" = "https://github.com/"* ]]`

File: Library/Homebrew/cmd/update.sh [hunk 13]
```diff
@@ -640,10 +631,28 @@ EOS
...
+          UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY_URL#https://github.com/}"
+          UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY%.git}"
+        fi
+        # HOMEBREW_GITHUB_API_TOKEN is optionally defined in the user environment.
+        # Global token HOMEBREW_GITHUB_API_TOKEN supersedes local defined one
+        # shellcheck disable=SC2153
+        if [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
+        then
+          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${HOMEBREW_GITHUB_API_TOKEN}")
+        elif [[ -n "${BASH_REMATCH[4]}" ]]
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/update.sh:650]: Ok, I'm happy to trust it to.

File: Library/Homebrew/cmd/update.sh [hunk 14]
```diff
@@ -640,10 +631,31 @@ EOS
...
+        then
+          UPSTREAM_REPOSITORY="${BASH_REMATCH[5]%.git}"
+        else
+          UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY_URL#https://github.com/}"
+          UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY%.git}"
+        fi
+        # HOMEBREW_GITHUB_API_TOKEN is optionally defined in the user environment.
+        # Global token HOMEBREW_GITHUB_API_TOKEN supersedes local defined one
+        # shellcheck disable=SC2153
+        if [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:647]:
```suggestion
        if [[ "${UPSTREAM_REPOSITORY_URL}" = "https://github.com/"* ]] && 
           [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
```

or similar as otherwise `HOMEBREW_GITHUB_API_TOKEN` being set at all may limit this.

It may be desirable to have a single `[[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]]` check which sets a variable e.g. `UPSTREAM_REPOSITORY_GITHUB_GLOB` or something which you can then check in these multiple places later.

File: Library/Homebrew/cmd/update.sh [hunk 15]
```diff
@@ -637,13 +628,36 @@ EOS
...
     (
       UPSTREAM_REPOSITORY_URL="$(git config remote.origin.url)"
+      unset UPSTREAM_REPOSITORY
+      unset UPSTREAM_REPOSITORY_TOKEN
 
       # HOMEBREW_UPDATE_FORCE and HOMEBREW_UPDATE_AUTO aren't modified here so ignore subshell warning.
       # shellcheck disable=SC2030
-      if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]]
+
+      if [[ "${UPSTREAM_REPOSITORY_URL}" = "https://github.com/"* ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:637]:
```suggestion
      if [[ "${UPSTREAM_REPOSITORY_URL}" = "https://github.com/"* ]]
```

File: Library/Homebrew/cmd/update.sh [hunk 16]
```diff
@@ -637,13 +628,36 @@ EOS
...
+      then
+        # HOMEBREW_GITHUB_API_TOKEN is optionally defined in the user environment.
+        # Global token HOMEBREW_GITHUB_API_TOKEN supersedes local defined one
+        # shellcheck disable=SC2153
+        if [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
+        then
+          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${HOMEBREW_GITHUB_API_TOKEN}")
+        elif [[ -n "${UPSTREAM_REPOSITORY_TOKEN}" ]]
+        then
+          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${UPSTREAM_REPOSITORY_TOKEN}")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:657]:
```suggestion
        if [[ -n "${UPSTREAM_REPOSITORY_TOKEN}" ]]
        then
          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${UPSTREAM_REPOSITORY_TOKEN}")
        elif [[ -n "${HOMEBREW_GITHUB_API_TOKEN}" ]]
        then
          CURL_GITHUB_API_ARGS=("--header" "Authorization: token ${HOMEBREW_GITHUB_API_TOKEN}")
```

**Comment by MikeMcQuaid** [comment]
@sazriel26 Looking better! Final two suggestions, will get one other maintainer to look this over and should be good to merge. Thanks for your persistence and patience, much appreciated ❤️

**Review by MikeMcQuaid — COMMENTED** [review]
One final tiny one-character change!

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me now, great work @sazriel26! Requesting another review.

**Review by carlocab — COMMENTED** [review]
Looks reasonable to me, but I want a little more time to mull this over to say for sure (given that it will be very hard to take back a broken `update.sh`).

Also: should we exclude Homebrew-owned repositories from these new code paths? One should never need a token for them, and it'll help make sure any corner-cases we've overlooked, if any, don't end up breaking `update.sh` in an unrecoverable way.

**Comment by carlocab** [comment]
Yes, agreed. Excluding Homebrew-owned repos _probably_ doesn't do anything. On the other hand, the cost of being wrong could be pretty bad.

**Comment by MikeMcQuaid** [comment]
> Also: should we exclude Homebrew-owned repositories from these new code paths?

@carlocab something like `"${DIR}" == "${HOMEBREW_REPOSITORY}"` should be sufficient as, if there's an issue here, a `brew update` of Homebrew/brew will fix it. Could you make a code suggestion of where you'd suggest that be placed?

File: Library/Homebrew/cmd/update.sh [hunk 17]
```diff
@@ -637,13 +628,35 @@ EOS
...
+      unset UPSTREAM_REPOSITORY
+      unset UPSTREAM_REPOSITORY_TOKEN
 
       # HOMEBREW_UPDATE_FORCE and HOMEBREW_UPDATE_AUTO aren't modified here so ignore subshell warning.
       # shellcheck disable=SC2030
       if [[ "${UPSTREAM_REPOSITORY_URL}" == "https://github.com/"* ]]
       then
         UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY_URL#https://github.com/}"
         UPSTREAM_REPOSITORY="${UPSTREAM_REPOSITORY%.git}"
+      elif [[ "${UPSTREAM_REPOSITORY_URL}" =~ https://(([^:@]+)(:([^@]+))?@)?github.com/(.*)$ ]]
```
Comment (Bo98) [inline Library/Homebrew/cmd/update.sh:640]:
```suggestion
      elif [[ "${UPSTREAM_REPOSITORY_URL}" =~ https://(([^:@/?#]+)(:([^@/?#]+))?@)github\.com/(.*)$ ]]
```

- Group 1 `?` seemed unnecessary as that just reduces it down to the earlier regex.
- `/` and similar URL component separators should not be matched in the username password as it's not valid there. Probably could be stricter but this is  good enough
- Escape `.` so we're only actually matching GitHub

**Review by Bo98 — APPROVED** [review]
Looks OK and quite safe to me with these suggested adjustments

File: Library/Homebrew/cmd/update.sh (hunk 17 above)
Comment (cho-m) [reply Library/Homebrew/cmd/update.sh:640]: > * Group 1 `?` seemed unnecessary as that just reduces it down to the earlier `if` condition.

Is Group 1 itself even necessary? `${BASH_REMATCH[1]}` isn't used so perhaps remove group and shift numbers down.

Comment (Bo98) [reply Library/Homebrew/cmd/update.sh:640]: Looks not actually

Comment (carlocab) [inline Library/Homebrew/cmd/update.sh:640]:
```suggestion
      elif [[ "${DIR}" != "${HOMEBREW_REPOSITORY}" ]] && [[ "${UPSTREAM_REPOSITORY_URL}" =~ https://(([^:@]+)(:([^@]+))?@)?github.com/(.*)$ ]]
```
I think this is what we want to do here based on discussion with @MikeMcQuaid.

**Review by carlocab — COMMENTED** [review]
I'm convinced that this code does what it's intended and is unlikely to break anything important.

However, is this the correct API for providing repository/tap-specific credentials?

IIUC, encoding access credentials in the URL is [insecure and discouraged](https://security.stackexchange.com/a/990). If so, then, at minimum, we should:

1. at least redact the credentials from the URL we pass to `curl`; or,
2. maybe consider a different API for passing these credentials.

**Comment by Bo98** [comment]
> redact the credentials from the URL we pass to curl

We're not passing a auth in the URL. We're deciding it from an already-encoded git URL and passing it to curl via `--header`.

On the credential retrieval side, it isn't fully foolproof. You can supply git with a plain URL and retrieve credentials via the keychain (where available) or use `.extraheader` (https://github.com/Homebrew/actions/blob/d54a6744d5fcdff54b45a9659f3e17f769389952/setup-homebrew/main.sh#L129). Though anything is better than nothing.

**Comment by MikeMcQuaid** [comment]
> IIUC, encoding access credentials in the URL is [insecure and discouraged](https://security.stackexchange.com/a/990).

In my experience: it's also the only method that 100% works across `curl` and `git` invocations cross-platform without other tooling and configuration.

That's not to say "let's encourage it" but it is a completely reasonable thing to do particularly given:

> We're not passing auth in the URL to curl. We're decoding it from an already-encoded git URL and passing it to curl via `--header`.

**Review by carlocab — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Thanks for your first PR to Homebrew, @sazriel26, you rock and I appreciate your patience with all the back-and-forth!

**Comment by p-linnane** [comment]
@sazriel26 Can you please rebase this so we can get it merged? Thanks!

**Comment by p-linnane** [comment]
You can check [the Git manual](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) for this.

**Comment by MikeMcQuaid** [comment]
> it is saying "Current branch parse-git-token is up to date."

You need to instead run `git fetch --all; git rebase origin/master` (assuming `origin` is Homebrew/brew and not your fork).



> Shall I reopen a new pull request from the parse-git-token (old master from which I have opened this PR)?

Yes, please.
